pfio/cache/multiprocess_file_cache.py

```diff
@@ -227,34 +231,42 @@ def _get(self, i):
 
         self._open_fds()
         offset = self.buflen * i
+
         fcntl.flock(self.cache_fd, fcntl.LOCK_SH)
-        index_entry = os.pread(self.cache_fd, self.buflen, offset)
-        (o, l) = unpack('Qq', index_entry)
-        if l < 0 or o < 0:
+        with record(f"pfio.cache.multiprocessfile:get:lock-{self.cache_fd}",
+                    trace=self.trace):
+            index_entry = os.pread(self.cache_fd, self.buflen, offset)
+            (o, l) = unpack('Qq', index_entry)
+            if l < 0 or o < 0:
+                fcntl.flock(self.cache_fd, fcntl.LOCK_UN)
+                return None
+
+            with record("pfio.cache.multiprocessfile:get:read",
+                        trace=self.trace):
+                data = os.pread(self.cache_fd, l, o)
+            assert len(data) == l
             fcntl.flock(self.cache_fd, fcntl.LOCK_UN)
-            return None
 
-        data = os.pread(self.cache_fd, l, o)
-        assert len(data) == l
-        fcntl.flock(self.cache_fd, fcntl.LOCK_UN)
         return data
 
     def put(self, i, data):
-        if self._frozen or self.closed:
-            return False
+        with record("pfio.cache.multiprocessfile:put", trace=self.trace):
+            if self._frozen or self.closed:
+                return False
 
-        try:
-            if self.do_pickle:
-                data = pickle.dumps(data)
-            return self._put(i, data)
+            try:
+                if self.do_pickle:
+                    data = pickle.dumps(data)
+                return self._put(i, data)
 
-        except OSError as ose:
-            # Disk full (ENOSPC) possibly by cache; just warn and keep running
-            if ose.errno == errno.ENOSPC:
-                warnings.warn(ose.strerror, RuntimeWarning)
-                return False
-            else:
-                raise ose
+            except OSError as ose:
+                # Disk full (ENOSPC) possibly by cache;
+                # just warn and keep running
+                if ose.errno == errno.ENOSPC:
+                    warnings.warn(ose.strerror, RuntimeWarning)
+                    return False
+                else:
+                    raise ose
 
     def _put(self, i, data):
```

Comment:
Like file_cache, we'd like to track I/O-related and lock operations separately. Separating them will be useful for diagnosing whether I/O or just lock wait is long.